cargo-annotation/annotation.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+import os
+import subprocess
+import json
+import sys
+import shlex
+
+
+def emit_annotation_for_clippy(results, with_annotation):
+    total_count = 0
+    limit = 10
+
+    severenty_map = {
+        "help": "notice",
+        "note": "notice",
+        "warning": "warning",
+        "error": "error",
+    }
+    for item in results:
+        if total_count >= limit:
+            break
+
+        message = item.get("message")
+        if not message:
+            continue
+
+        spans = message.get("spans") or []
+        primary_span = next((span for span in spans if span.get("is_primary")), None)
+        if not primary_span:
+            continue
+
+        level = severenty_map.get(message.get("level"), "error")
+        title = message.get("message", "")
+        rendered_message = message.get("rendered", "")
+
+        file_name = (primary_span["file_name"],)
+        line_start = (primary_span["line_start"],)
+        line_end = (primary_span["line_end"],)
+        column_start = (primary_span["column_start"],)
+        column_end = (primary_span["column_end"],)
+
+        line_info = f"line={line_start},endLine={line_end},title={title}"
+
+        column_info = ""
+        if (
+            line_start == line_end
+            and column_end is not None
+            and column_start is not None
+        ):
+            column_info = f"col={column_start},endColumn={column_end},"
+        if with_annotation:
+            print(
+                f"::{level} file={file_name},{column_info}{line_info}::{rendered_message}"
+            )
+        total_count += 1
+
+    return 1 if total_count > 0 else 0
+
+
+def main():
+    cmd = ["cargo"]
+
+    cargo_command = os.getenv("INPUT_CARGO_COMMAND", "").strip()
+    with_annotation = (
+        os.getenv("INPUT_WITH_ANNOTATION", "true").strip().lower() == "true"
+    )
+
+    if cargo_command:
+        args = shlex.split(cargo_command)
+        cmd.extend(args)
+
+    if "--message-format=json" not in cmd:
+        cmd.append("--message-format=json")
+
+    proc = subprocess.run(cmd, capture_output=with_annotation)
+    clippy_results = [
+        json.loads(line) for line in proc.stdout.splitlines() if line.strip()
+    ]
+
+    result = emit_annotation_for_clippy(clippy_results, with_annotation)
+    sys.exit(result)
```

Comment:
I think you can simplify this a bit, by only parsing annotations when `with_annotation` is true and then returning the processes return code:

```suggestion
    if with_annotation:
        emit_annotation_for_clippy(clippy_results)
    sys.exit(proc.returncode)
```

Reply:
Yeah, clippy will not return a failure code unless `-D warning` is passed, but then that will cause clippy to error just after the first warning. It is more useful in the with_annotation=true case to not rely on the exit code (we can show up to 10 warnings), so to me it makes sense to use that same logic even when GH annotation is turned off.

Reply:
Makes sense. Thank you!